src/qililab/instruments/qblox/qblox_module.py

```diff
@@ -551,6 +551,15 @@ def reset(self):
         self.clear_cache()
         self.device.reset()
 
+    def upload_qpysequence(self, qpysequence: QpySequence, port: str):
+        """Upload the qpysequence to its corresponding sequencers."""
+        sequencers = self.get_sequencers_from_chip_port_id(chip_port_id=port)
+        for sequencer in sequencers:
+            logger.info("Sequence program: \n %s", repr(qpysequence._program))  # pylint: disable=protected-access
+            self.device.sequencers[sequencer.identifier].sequence(qpysequence.todict())
+            self.device.sequencers[sequencer.identifier].sync_en(True)
+            self.sequences[sequencer.identifier] = (qpysequence, True)
+
```

Comment:
would it be interesting to keep the `if not uploaded` logic as in the `upload` method below? Otherwise there's not much sense on keeping the tuple `(sequence: qpysequence,uploaded: bool)` is there?

Reply:
Nope, the thing is that I want to bypass the caching dictionary that `qblox_module` is using here. Every time I call `upload` I want the provided `qpysequence` to be uploaded without checking any cache. This is another argument that `compile` should be done outside the `qblox_module`. (since the original `upload` uses results from the `compile` method)

Reply:
Doesn't that make execution slower though if we have to re-upload every time even if the sequence is the same?

Reply:
But everytime you run a QProgram, you compile again and upload again, it's not like the compilation of software-loops with Qibo circuit, that you may have the same circuit between iterations.

Reply:
the current cache for `qblox_module` (qblox compiler in the current workflow) we check whether the pulse schedule is the same or not (note that pulse schedule might not necessarily come from qibo circuits, it could be a custom pulse schedule)
at `qblox_module.compile`:
```python
        compiled_sequences = []
        sequencers = self.get_sequencers_from_chip_port_id(chip_port_id=pulse_bus_schedule.port)
        for sequencer in sequencers:
            if pulse_bus_schedule != self._cache.get(sequencer.identifier):
                sequence = self._compile(pulse_bus_schedule, sequencer)
                compiled_sequences.append(sequence)
            else:
                compiled_sequences.append(self.sequences[sequencer.identifier][0])
        return compiled_sequences
```
I don't know if a similar thing could be done here and I just realize this is done before even compiling so it might save some time to save compiled pulse schedules and check for repetitions at sequential executions of a program. 
But maybe that should be a separate task?

Reply:
But still the cache is about `PulseBusSchedule`, not `qpysequence`:

```
_cache: dict[int, PulseBusSchedule] = {}
```

And I am certain that this is the wrong place to have the cache, and the `compile` method too. The instrument class should just do `upload`, `run`, `get_acquisitions` and similar tasks. The orchestration of the whole flow should be in a much higher level like `platform`.

`QbloxModule` should not know that `PulseBusSchedule` exists, in the same sense that it does not know that a `QProgram` exists.

The weird syntax in `upload_qpysequence`: `self.sequences[sequencer.identifier] = (qpysequence, True)`  is because this line in `__init__` exists and I don't want to change it, since it will break the Qibo flow.

```
# The sequences dictionary contains all the compiled sequences for each sequencer and a flag indicating whether
# the sequence has been uploaded or not
self.sequences: dict[int, tuple[QpySequence, bool]] = {}  # {sequencer_idx: (program, True), ...}
```

Reply:
`self.sequences: dict[int, tuple[QpySequence, bool]] = {}`  should become `self.sequences: dict[int, QpySequence] = {}` in the alternative implementation that I am describing.

Reply:
We can create a separate task to investigate this and discuss with the team.